src/Application.php

```diff
@@ -85,212 +89,165 @@ class Application extends MiddlewarePipe
-        $this->router          = $router;
-        $this->container       = $container;
-        $this->defaultDelegate = $defaultDelegate;
-        $this->emitter         = $emitter;
+        $this->pipeline          = new MiddlewarePipe();
+        $this->router            = $router;
+        $this->container         = $container;
+        $this->middlewareFactory = new MiddlewareFactory($container);
+        $this->defaultHandler    = $defaultHandler;
+        $this->emitter           = $emitter;
         $this->responsePrototype = $responsePrototype ?: new Response();
     }
 
-    /**
-     * @param callable|string $middleware Middleware (or middleware service name) to associate with route.
-     * @param null|string $name The name of the route.
-     */
-    public function get(string $path, $middleware, string $name = null) : Router\Route
+    public function handle(ServerRequestInterface $request) : ResponseInterface
     {
-        return $this->route($path, $middleware, ['GET'], $name);
+        return $this->process($request, $this->getDefaultHandler());
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler) : ResponseInterface
+    {
+        return $this->pipeline->process($request, $handler);
     }
 
     /**
-     * @param callable|string $middleware Middleware (or middleware service name) to associate with route.
-     * @param null|string $name The name of the route.
+     * Run the application
+     *
+     * If no request or response are provided, the method will use
+     * ServerRequestFactory::fromGlobals to create a request instance, and
+     * instantiate a default response instance.
+     *
+     * It retrieves the default delegate using getDefaultHandler(), and
+     * uses that to process itself.
+     *
+     * Once it has processed itself, it emits the returned response using the
+     * composed emitter.
      */
-    public function post(string $path, $middleware, $name = null) : Router\Route
+    public function run(ServerRequestInterface $request = null, ResponseInterface $response = null) : void
     {
-        return $this->route($path, $middleware, ['POST'], $name);
+        try {
+            $request  = $request ?: ServerRequestFactory::fromGlobals();
+        } catch (InvalidArgumentException | UnexpectedValueException $e) {
+            // Unable to parse uploaded files | Invalid request method
+            $this->emitMarshalServerRequestException($e);
+            return;
+        }
+
+        $response = $this->handle(
+            $request->withAttribute('originalResponse', $response ?: new Response())
+        );
+
+        $emitter = $this->getEmitter();
+        $emitter->emit($response);
     }
 
     /**
-     * @param callable|string $middleware Middleware (or middleware service name) to associate with route.
-     * @param null|string $name The name of the route.
+     * @param string|array|callable|MiddlewareInterface $middleware Middleware
```

Comment:
As I've suggested in another place order of types - `null|(simple types in alphabetical order)|(complex types in alphabetical order)` - so here we should have `array|callable|string|MiddlewareInterface`.

Reply:
I would argue for the following order:

- `null`
- all other scalar types, in alphabetical order
- complex built-in types (`array`, `callable`, `resource`) in alphabetical order
- instance types, in alphabetical order

So, clearly we have differing opinions, and we'll need to make a decision. In the meantime, this is not the main point of this PR at this time; I need feedback on _architecture_ and _design_.